frontend/src/pages/projects/screens/spawner/service.ts

```diff
@@ -60,42 +123,127 @@ export const createConfigMapsAndSecretsForNotebook = async (
+): Promise<EnvironmentFromVariable[]> => {
+  const creatingPromises = getPromisesForConfigMapsAndSecrets(projectName, envVariables, 'create');
+
   return Promise.all(creatingPromises)
-    .then((results: (ConfigMapKind | SecretKind)[]) => {
-      return results.reduce<EnvironmentFromVariable[]>((acc, resource) => {
-        let envFrom;
-        if (resource.kind === 'Secret') {
-          envFrom = {
-            secretRef: {
-              name: resource.metadata.name,
-            },
-          };
-        } else if (resource.kind === 'ConfigMap') {
-          envFrom = {
-            configMapRef: {
-              name: resource.metadata.name,
-            },
-          };
-        } else {
-          return acc;
-        }
-        return [...acc, envFrom];
-      }, []);
-    })
+    .then((results: (ConfigMapKind | SecretKind)[]) => getEnvFromList(results, []))
     .catch((e) => {
       console.error('Creating environment variables failed: ', e);
       throw e;
     });
 };
+
+export const updateConfigMapsAndSecretsForNotebook = async (
+  projectName: string,
+  notebook: NotebookKind,
+  envVariables: EnvVariable[],
+): Promise<EnvironmentFromVariable[]> => {
+  const existingEnvVars = await fetchNotebookEnvVariables(notebook);
+  const [oldResources, newResources] = _.partition(envVariables, (envVar) => envVar.existingName);
+  const currentNames = oldResources
+    .map((envVar) => envVar.existingName)
+    .filter((v): v is string => !!v);
+  const removeResources = existingEnvVars.filter(
+    (envVar) => envVar.existingName && !currentNames.includes(envVar.existingName),
+  );
+
+  const [typeChangeResources, updateResources] = _.partition(oldResources, (envVar) =>
+    existingEnvVars.find(
+      (existingEnvVar) =>
+        existingEnvVar.existingName === envVar.existingName &&
+        existingEnvVar.values?.category !== envVar.values?.category,
+    ),
+  );
+
+  const deleteResources = [...removeResources, ...typeChangeResources];
+  // will only delete generic files here when updating because we only map them
+  const deletingPromises = deleteResources
+    .map((envVar) => {
+      if (!envVar.existingName) {
+        return null;
+      }
+      switch (envVar.values?.category) {
+        case SecretCategory.GENERIC:
+          return deleteSecret(projectName, envVar.existingName);
+        case ConfigMapCategory.GENERIC:
+          return deleteConfigMap(projectName, envVar.existingName);
+        default:
+          return null;
+      }
+    })
+    .filter((v): v is Promise<K8sStatus> => !!v);
+  const creatingPromises = getPromisesForConfigMapsAndSecrets(
+    projectName,
+    [...newResources, ...typeChangeResources],
+    'create',
+  );
+  const updatingPromises = getPromisesForConfigMapsAndSecrets(
+    projectName,
+    updateResources,
+    'update',
+  );
+
+  const [created] = await Promise.all([
+    Promise.all(creatingPromises),
+    Promise.all(deletingPromises),
+    Promise.all(updatingPromises),
```

Comment:
Spread these, no need to double compound promise.all... I think you're reversing the array by doing destructring.

No harm, but a little weird to read.